src/main/java/com/sap/sgs/phosphor/fosstars/model/score/WeightedCompositeScore.java

```diff
@@ -28,84 +29,82 @@
 public class WeightedCompositeScore extends AbstractScore implements Tunable {
 
   /**
-   * The default weight for sub-scores.
+   * A set of weighted sub-scores.
    */
-  private static final double DEFAULT_WEIGHT = 1.0;
+  private final Set<Score> subScores;
 
   /**
-   * A set of weighted sub-scores.
+   * Weights of the sub-scores.
    */
-  private final Set<WeightedScore> weightedScores;
+  private final ScoreWeights weights;
 
   /**
    * Makes a score from a list of sub-scores.
    */
-  public WeightedCompositeScore(String name, Score... scores) {
-    this(name, prepare(scores));
+  public WeightedCompositeScore(String name, Score... subScores) {
+    this(name, setOf(subScores), ScoreWeights.createFor(subScores));
   }
 
   /**
-   * This is a constructor which is used by Jackson during deserialization.
-   * TODO: Check if Jackson changes access for this constructor, and don't revert it back.
-   *       If so, is it a security issue?
+   * Initializes a new score.
+   * This constructor is used by Jackson during deserialization.
    */
   @JsonCreator
   WeightedCompositeScore(
       @JsonProperty("name") String name,
-      @JsonProperty("weightedScores") Set<WeightedScore> weightedScores) {
+      @JsonProperty("subScores") Set<Score> subScores,
+      @JsonProperty("weights") ScoreWeights weights) {
 
     super(name);
-    this.weightedScores = check(weightedScores);
+
+    Objects.requireNonNull(subScores, "Sub-scores can't be null!");
+    Objects.requireNonNull(weights, "Weights can't be null");
+
+    if (subScores.isEmpty()) {
+      throw new IllegalArgumentException("Sub-scores can't be empty!");
+    }
+
+    if (subScores.size() != weights.size()) {
+      throw new IllegalArgumentException(
+          "Hey! Number of sub-scores should be equal to number of weights!");
+    }
+
+    for (Score score : subScores) {
+      if (!weights.of(score).isPresent()) {
```

Comment:
A question here:
- Should `weight = weights.of(score)` contain in `weights` as well?